packages/kbn-data-stream-adapter/src/data_stream_adapter.ts

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import type {
+  ClusterPutComponentTemplateRequest,
+  IndicesIndexSettings,
+  IndicesPutIndexTemplateIndexTemplateMapping,
+  IndicesPutIndexTemplateRequest,
+} from '@elastic/elasticsearch/lib/api/types';
+import type { Logger, ElasticsearchClient } from '@kbn/core/server';
+import type { Subject } from 'rxjs';
+import type { FieldMap } from './field_maps/types';
+import { createOrUpdateComponentTemplate } from './create_or_update_component_template';
+import { createOrUpdateDataStream } from './create_or_update_data_stream';
+import { createOrUpdateIndexTemplate } from './create_or_update_index_template';
+import { InstallShutdownError, installWithTimeout } from './install_with_timeout';
+import { getComponentTemplate, getIndexTemplate } from './resource_installer_utils';
+
+export interface DataStreamAdapterParams {
+  kibanaVersion: string;
+  totalFieldsLimit?: number;
+}
+export interface SetComponentTemplateParams {
+  name: string;
+  fieldMap: FieldMap;
+  settings?: IndicesIndexSettings;
+  dynamic?: 'strict' | boolean;
+}
+export interface SetIndexTemplateParams {
+  name: string;
+  componentTemplateRefs?: string[];
+  namespace?: string;
+  template?: IndicesPutIndexTemplateIndexTemplateMapping;
+  hidden?: boolean;
+}
+
+export interface GetInstallFnParams {
+  logger: Logger;
+  pluginStop$: Subject<void>;
+  tasksTimeoutMs?: number;
+}
+export interface InstallParams {
+  logger: Logger;
+  esClient: ElasticsearchClient;
```

Comment:
To provide more flexibility in usage of the install function, can we have
`esClient: Promise<ElasticsearchClient>;`